File: examples/klein-rational-expectations/klein-rational-expectations.R
#!/usr/bin/env Rscript

# Solve the public forward-looking Klein example with BIMETS R.
#
# Source: model and terminal-value experiment are from the BIMETS R SIMULATE
# example and the original public repository:
# https://github.com/andrea-luciani/bimets#rational-expectations.

suppressPackageStartupMessages(library(bimets))
script_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
example_dir <- dirname(normalizePath(sub("^--file=", "", script_arg[[1]])))
model_text <- paste(readLines(file.path(example_dir, "klein-rational-expectations.mdl"), warn = FALSE), collapse = "\n")
model <- LOAD_MODEL(modelText = model_text, quietly = TRUE)
model <- LOAD_MODEL_DATA(model, CSV2BIMETS(file.path(example_dir, "klein-data.csv"), mergedList = TRUE), quietly = TRUE)
model$modelData$i[[1931, 1]] <- 2
model <- ESTIMATE(model, quietly = TRUE)
model <- SIMULATE(model, simAlgo = "NEWTON", simConvergence = 1e-6, simIterLimit = 200, TSRANGE = c(1924, 1, 1930, 1), quietly = TRUE)

values <- as.numeric(window(model$simulation$i, start = c(1924, 1), end = c(1930, 1)))
cat("year,i\n")
cat(sprintf("%d,%.6f\n", 1923 + seq_along(values), values), sep = "")

File: examples/klein-rational-expectations/klein-rational-expectations.py
#!/usr/bin/env python3
"""Solve the public forward-looking Klein example with bimets Python.

Source
------
The model and terminal-value experiment come from the BIMETS R ``SIMULATE``
example and the original public repository:
https://github.com/andrea-luciani/bimets#rational-expectations.
"""

from pathlib import Path

from bimets import CSV2BIMETS, ESTIMATE, LOAD_MODEL, SIMULATE, BimetsDataset

EXAMPLE_DIR = Path(__file__).resolve().parent


def main() -> None:
    """Estimate the behaviorals and solve investment over 1924--1930."""
    model = LOAD_MODEL(
        model_text=(EXAMPLE_DIR / "klein-rational-expectations.mdl").read_text(
            encoding="utf-8"
        )
    )
    data = BimetsDataset(
        CSV2BIMETS(EXAMPLE_DIR / "klein-data.csv", merged=True)
    ).assign_range({"i": 2.0}, start=(1931, 1))
    estimates = ESTIMATE(model, data)
    result = SIMULATE(
        model,
        data,
        coefficients=estimates,
        time_range=(1924, 1, 1930, 1),
        algorithm="NEWTON",
        convergence=1e-6,
        max_iterations=200,
    )

    print("year,i")
    output = result["i"][[[1924, 1], [1930, 1]]]
    for position, value in enumerate(output.values):
        print(f"{1924 + position},{value:.6f}")


if __name__ == "__main__":
    main()
